### PR TITLE
Fix imaging value for generating sysimages for multiple targets

### PR DESCRIPTION
**Pull request by ven-k**
Fixes `imaging_default` function, so that sysimages for multiple architectures can be built without setting `pkgimages=yes`

### Current issue:
For versions >1.9-dev,
For a use-case like:
```julia
julia> sysimage= unsafe_string(Base.JLOptions().image_file)

julia>  run(`./julia
               --sysimage=$sysimage
               --cpu-target='generic;sandybridge,-xsaveopt,clone_all;haswell,-rdrnd,base(1)'
               --output-o=mock.o mock.jl
               --pkgimages=no`)

fatal: error thrown and no exception handler available.
ErrorException("More than one command line CPU targets specified without a `--output-` flag specified")

[30365] signal (11.1): Segmentation fault
in expression starting at none:0
```

This is because, with `pkgimages=no`, even if output-o is passed, the `imaging_default` returns false. This in-turn causes issues while generating sysimages for multiple cpu-targets

### After this PR, these work:
1. Sysimages with multiple targets and no pkgimages
```
julia> run(`julia --sysimage=$sysimage --cpu-target='generic;sandybridge,-xsaveopt,clone_all;haswell,-rdrnd,base(1)' --output-o=mock.o mock.jl --pkgimages=no`)
```
2. Pkgimages=yes and multiple targets.
   From the [docs](https://docs.julialang.org/en/v1.9-dev/devdocs/pkgimg/#Package-images-optimized-for-multiple-microarchitectures), there should be a unified cache. It's unclear to me if the current change results an issue in that regard. An alternative would be changing to `return jl_options.image_codegen || jl_generating_output()`.
```
julia> run(`julia --cpu-target='generic;sandybridge,-xsaveopt,clone_all;haswell,-rdrnd,base(1)' --pkgimages=yes`)
```

### CODE REVIEW

**Comment by vchuravy** [comment]
Could you add a test here?

File: test/cmdlineargs.jl
Comment (vchuravy) [inline test/cmdlineargs.jl:601]: This probably needs a guard for x86_64 or a different list for supported architectures.

File: test/cmdlineargs.jl
Comment (vchuravy) [inline test/cmdlineargs.jl:611]: Okay this should be propagated to the target process. https://github.com/JuliaLang/julia/blob/a23b29cb1a6013277e381092147fc527a44d924a/base/util.jl#L171

As should `JULIA_CPU_TARGET`. https://github.com/JuliaLang/julia/blob/a23b29cb1a6013277e381092147fc527a44d924a/base/loading.jl#L2071

File: test/cmdlineargs.jl
Comment (vtjnash) [inline test/cmdlineargs.jl:602]:
```suggestion
        --cpu-target='generic;generic'
```

File: test/cmdlineargs.jl
Comment (vtjnash) [inline test/cmdlineargs.jl:626]: this test sounds slow, and should not be included without further reconsideration of performance

Comment (ven-k) [reply test/cmdlineargs.jl:626]: Yes. I've noticed that this test adds ~160s. It is too much time for 1 single test.
So, for easy dropping, I've added it in a dedicated commit.

I couldn't think of any other simpler test for output-o, which is faster, though.

Comment (ven-k) [reply test/cmdlineargs.jl:626]: @vtjnash How about this: we intentionally don't pass the sysimage arg. When it fails we know everything before that passed, including our case. This takes only 0.05s

```
# Sysimage with multiple cpu targets
@testset "Sysimage for multiple microarchitectures" begin
    julia_path = joinpath(Sys.BINDIR, Base.julia_exename())
    outputo_file = tempname()
    write(outputo_file, "1")
    object_file = tempname() * ".o"

    # This is to test that even with `pkgimages=no`, we can create object file
    # with multiple cpu-targets
    # The cmd is checked for --object-o as soon as it is run. So, to avoid long
    # testing times, intentionally don't pass sysimage; when we reach the
    # corresponding error, we know that `check_cmdline` has already passed
    let v = readchomperrors(`$julia_path
        --cpu-target='native;generic'
        --output-o=$object_file $outputo_file
        --pkgimages=no`)

        @test v[1] == false
        @test v[2] == ""

        err = first(split(v[3], "\n\n"))
        @test err != "fatal: error thrown and no exception handler available.\nErrorException(\"More than one command line CPU targets specified without a `--output-` flag specified\")"
        @test err == "ERROR: File \"boot.jl\" not found"
    end
end
```

**Comment by DilumAluthge** [comment]
This PR is marked as `merge me`, but it seems that there are still some unresolved conversations. So I'll remove the label.

@vtjnash or @vchuravy Can you add the `merge me` label back once this PR is ready?

**Comment by DilumAluthge** [comment]
Also, looks like the PowerPC build is broken @vchuravy

File: test/cmdlineargs.jl
Comment (vtjnash) [reply test/cmdlineargs.jl:626]: yes, though let's do `@test !contains(err, "More than one command line CPU targets specified")` so it will still work if other minor aspects change (though the second test for `==` will also catch those)

**Comment by DilumAluthge** [comment]
@vchuravy Do you want me to mark the PowerPC build as "allow fail"?

(The PowerPC tests are already marked as "allow fail", but currently the PowerPC build is marked as "regular".)

**Comment by vchuravy** [comment]
No the PPC build is not allowed to fail.

**Comment by DilumAluthge** [comment]
It seems like this PR is a release blocker for 1.9. Are we willing to let a broken PowerPC build block this PR? The website says that PowerPC is a Tier 3 platform.

**Comment by vchuravy** [comment]
Reproducer is:

```
julia> cd("src") do
       end
```

Which lead to:

```
julia> function my_cd(dir::AbstractString)                                                                                                                                                                                                                                                                                    
         fd = ccall(:open, Int32, (Cstring, Int32), :., 0)                                                                                                                                                                                                                                                                    
         cd(dir)                                                                                                                                                                                                                                                                                                              
       end                                                                                                                                                                                                                                                                                                                    
my_cd (generic function with 1 method)                                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                              
julia> my_cd("src")                                                                                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                                                                              
[4077090] signal (11.1): Segmentation fault                                                                                                                                                                                                                                                                                   
in expression starting at REPL[4]:1                                                                                                                                                                                                                                                                                           
sizeof at ./Base.jl:217 [inlined]                                                                                                                                                                                                                                                                                             
unsafe_convert at ./c.jl:216 [inlined]                                                                                                                                                                                                                                                                                        
cd at ./file.jl:90                   
```

The `ccall` to open is necessary.

Inlining `cd` still yields the same:

```
julia> function my_cd(dir::AbstractString)
         fd = ccall(:open, Int32, (Cstring, Int32), :., 0)
         systemerror(:open, fd == -1)
         err = ccall(:uv_chdir, Cint, (Cstring,), dir)
         return err
       end
```

Pretty minimal:

```
julia> function my_cd(dir::AbstractString)
         fd = ccall(:open, Int32, (Cstring, Int32), :., 0)
         systemerror(:open, fd == -1)
         dir = Base.cconvert(Cstring, dir)
         dir = Base.unsafe_convert(Cstring, dir)
         return dir
       end
```
Now why the hell does this PR break things...

**Comment by gbaraldi** [comment]
If you turn off package images does it change something?

**Comment by vchuravy** [comment]
> If you turn off package images does it change something?

Good catch `--pkgimages=no` makes it work...

**Comment by vtjnash** [comment]
`open` is vararg in C, so that ccall signature is illegal for PPC

**Comment by vchuravy** [comment]
That's a good observation. This PR also seems to change JIT compliation when `--pkgimages=yes`

```
 fence syncscope("singlethread") seq_cst                                                                                                                                                                                                                                                                                     
;  @ REPL[1]:1 within `my_cd`                                                                                                                                                                                                                                                                                                 
  %11 = load volatile i64, i64* %safepoint, align 8                                                                                                                                                                                                                                                                           
  fence syncscope("singlethread") seq_cst                                                                                                                                                                                                                                                                                     
;  @ REPL[1]:2 within `my_cd`                                                                                                                                                                                                                                                                                                 
; ┌ @ c.jl:234 within `unsafe_convert` @ pointer.jl:58                                                                                                                                                                                                                                                                        
   %12 = load {}**, {}*** bitcast ({}** @"jl_sym#.#2" to {}***), align 8                                                                                                                                                                                                                                                      
   %13 = getelementptr inbounds {}*, {}** %12, i64 3
   %14 = ptrtoint {}** %13 to i64
; └
  %15 = load atomic i32 (i64, i32)*, i32 (i64, i32)** bitcast (void ()** @jlplt_open_176_got to i32 (i64, i32)**) unordered, align 8
  %16 = call i32 %15(i64 %14, i32 0)
```

versus

```
  fence syncscope("singlethread") seq_cst                                                                                                                                                                                                                                                                                     
;  @ REPL[1]:1 within `my_cd`                                                                                                                                                                                                                                                                                                 
  %11 = load volatile i64, i64* %safepoint, align 8                                                                                                                                                                                                                                                                           
  fence syncscope("singlethread") seq_cst                                                                                                                                                                                                                                                                                     
;  @ REPL[1]:2 within `my_cd`                                                                                                                                                                                                                                                                                                 
  %12 = call i32 inttoptr (i64 140735334611648 to i32 (i64, i32)*)(i64 140735201562464, i32 0)                                                                                                                                                                                                                                
;  @ REPL[1]:4 within `my_cd`                                                                                                                                                                                                                                                                                                 
; ┌ @ c.jl:215 within `unsafe_convert` @ pointer.jl:60                                                                    
```

File: src/jitlayers.h
Comment (vchuravy) [inline src/jitlayers.h:76]: So this seems undesirable. It effectively causes codegen to use image mode by default even for normal JIT compilation.

Comment (vchuravy) [reply src/jitlayers.h:76]: The flag used to be:
```
static inline bool imaging_default() {
    return jl_options.image_codegen || (jl_generating_output() && !jl_options.incremental);
}
```

@vtjnash do we want it to be

```
jl_options.image_codegen || (jl_generating_output() && (!jl_options.incremental || jl_options.use_pkgimages))
```
???

Comment (vtjnash) [reply src/jitlayers.h:76]: Sure. I’d also be okay with `return 1` here xD

Comment (vchuravy) [reply src/jitlayers.h:76]: Well that would be a slightly more drastic change, but @pchintalapudi might like that as well.

We should evaluate for 1.10 if that is feasible

**Comment by staticfloat** [comment]
@ven-k can you please confirm that this PR fixes your original issue?

**Review by vchuravy — APPROVED** [review]
LGTM, modulo checking that we are not using imaging mode in normal JIT mode.

**Comment by ven-k** [comment]
With this I can pass `--cpu-targets` and generate target-compatible sysimages.

----
Note that when it is added to env via JULIA_CPU_TARGET  (Currently PackageCompiler does that for >=v1.9), on relocating I got a segfault. (But considering passing the `--cpu-targets` works, not a blocking issue for my use case)
```
./julia-fc4b079f69/bin/julia -J fc4b079f69_8.so

[5504] signal (7.2): Bus error
in expression starting at none:0
unknown function (ip: 0x7fb2ea3c75d3)
unknown function (ip: 0x7fb2ea3ab428)
unknown function (ip: 0x7fb2ea3ae61a)
unknown function (ip: 0x7fb2ea3b9d46)
_dl_catch_exception at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x7fb2ea3b9609)
unknown function (ip: 0x7fb2ea39934b)
_dl_catch_exception at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
_dl_catch_error at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x7fb2ea399b58)
dlopen at /lib/x86_64-linux-gnu/libdl.so.2 (unknown line)
ijl_load_dynamic_library at /cache/build/default-amdci5-0/julialang/julia-master/src/dlload.c:355
ijl_preload_sysimg_so at /cache/build/default-amdci5-0/julialang/julia-master/src/staticdata.c:2641
_finish_julia_init at /cache/build/default-amdci5-0/julialang/julia-master/src/init.c:827
julia_init at /cache/build/default-amdci5-0/julialang/julia-master/src/init.c:819
jl_repl_entrypoint at /cache/build/default-amdci5-0/julialang/julia-master/src/jlapi.c:711
main at /cache/build/default-amdci5-0/julialang/julia-master/cli/loader_exe.c:58
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x401098)
Allocations: 0 (Pool: 0; Big: 0); GC: 0

[5504] signal (7.2): Bus error
in expression starting at none:0
unknown function (ip: 0x7fb2ea3c75d3)
unknown function (ip: 0x7fb2ea3ab428)
unknown function (ip: 0x7fb2ea3ae61a)
unknown function (ip: 0x7fb2ea3b9d46)
_dl_catch_exception at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x7fb2ea3b9609)
unknown function (ip: 0x7fb2ea39934b)
_dl_catch_exception at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
_dl_catch_error at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x7fb2ea399b58)
dlopen at /lib/x86_64-linux-gnu/libdl.so.2 (unknown line)
ijl_load_dynamic_library at /cache/build/default-amdci5-0/julialang/julia-master/src/dlload.c:355
ijl_preload_sysimg_so at /cache/build/default-amdci5-0/julialang/julia-master/src/staticdata.c:2641
_finish_julia_init at /cache/build/default-amdci5-0/julialang/julia-master/src/init.c:827
julia_init at /cache/build/default-amdci5-0/julialang/julia-master/src/init.c:819
jl_repl_entrypoint at /cache/build/default-amdci5-0/julialang/julia-master/src/jlapi.c:711
main at /cache/build/default-amdci5-0/julialang/julia-master/cli/loader_exe.c:58
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x401098)
Allocations: 0 (Pool: 0; Big: 0); GC: 0
Bus error (core dumped)
```

**Comment by DilumAluthge** [comment]
@vchuravy @staticfloat Would it be possible to make `JULIA_CPU_TARGET` support the same values as `--cpu-targets`?

**Pull request merged**
